src/main/java/warehouse/Cart.java

```diff
@@ -24,8 +24,16 @@ public void addProductInCart(Product product) {
         devices.add(product);
     }
 
-    //METODO CHE AGGIUNGE AL CARRELLO TRAMITE ID
-    public void addToCartById(FilterById filterById, Warehouse warehouse) throws InvalidIdException {
+    //AGGIUNGO AL CARRELLO TRAMITE ID
+    public void addToCartById(FilterById filterById, Warehouse warehouse) throws InvalidIdException, EmptyWarehouseException {
+        for (Product product : devices) {
+            if (filterById.isInFilter(product)) {
+                throw new InvalidIdException("You have already add this product!");
```

Comment:
tornerei il messaggio "this id already exist in Cart" nell'exceptionn e come dicevamo in call implementerei questo controllo in un metodo ceh torna un bollano che chiami qua all'inizio